{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1ebb1721",
   "metadata": {},
   "outputs": [],
   "source": [
    "# packages to import, read in the latest dataframe pickle\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from matplotlib import pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "import matplotlib\n",
    "from colour import Color\n",
    "from glob import glob\n",
    "import re\n",
    "import PIL\n",
    "from PIL import Image\n",
    "\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "\n",
    "font = {'family' : 'sans-serif',\n",
    "        'weight' : 'normal',\n",
    "        'size'   : 14}\n",
    "\n",
    "matplotlib.rc('font', **font)\n",
    "\n",
    "pd.options.mode.chained_assignment = None "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "77d3a36b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this makes a custom colormap for visualizing the footprints\n",
    "#you input a list of colors (html codes), and a corresponding list of sizes\n",
    "#this will generate a colormap which will have the given color at the corresponding size, and then will interpolate colors to the next size\n",
    "\n",
    "def make_cp(colors, sizes):\n",
    "    cp =[tuple([0,0,0,0])]*sizes[0] #0-20\n",
    "    for i in range(len(colors)-1):\n",
    "        cs=Color(colors[i])\n",
    "        ce=Color(colors[i+1])\n",
    "        for j in list(cs.range_to(ce, sizes[i+1]-sizes[i])):\n",
    "            j=list(j.rgb)\n",
    "            j.append(1)\n",
    "            j=tuple(j)\n",
    "            cp.append(j)\n",
    "\n",
    "    return matplotlib.colors.ListedColormap(cp)\n",
    "\n",
    "#the colormap I use in general\n",
    "\n",
    "#colors Pol II and PIC sized footprints\n",
    "colors=['#ed306f','#e42cf6','#f3aafa','#265de6','#33a040'] \n",
    "sizes=[20,            45,        55,   70,         100]\n",
    "pol_cp=make_cp(colors, sizes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ce2aa39b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this makes a dataframe used to plot enrichment of specific footprint sizes w.r.t. position\n",
    "#s is the start coordinate, e is the end coordinate\n",
    "#min_s, max_s are the min and max size\n",
    "#bin_size is if you want to do bins of sizes (so like every 5nt or whatever)\n",
    "\n",
    "def plot_meta_heatmap(reads_df,s,e, min_s,max_s,bin_size):\n",
    "    m_d=pd.DataFrame()\n",
    "    reads_df=reads_df.fillna(0)[np.arange(s,e).astype(str)]\n",
    "    reads_df[reads_df<=2]=0\n",
    "    d=[]\n",
    "    for s in range(min_s//bin_size,max_s//bin_size):\n",
    "        df=reads_df.copy()\n",
    "        df=df.where(df >= (s)*bin_size,0)\n",
    "        df=df.where(df < (s+1)*bin_size,0)\n",
    "        a=pd.DataFrame(df.sum(axis=0)).T\n",
    "        d.append((s*bin_size))\n",
    "        m_d=pd.concat([m_d,a], axis=0)\n",
    "    m_d.index=d\n",
    "    return m_d"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "771468ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "#this generates an array with the enrichment of a given size range of footprints in a given range, normalized by read count at the position\n",
    "\n",
    "def meta_line_plot(reads_df, start, end, fp_size_min, fp_size_max):\n",
    "    \n",
    "    df=reads_df.copy()\n",
    "    \n",
    "    count_df=df.copy()\n",
    "    count_df[count_df!=2]=1\n",
    "    count_df[count_df==2]=0\n",
    "    counts=count_df.sum(axis=0)\n",
    "    counts=counts.loc[np.arange(start,end).astype(str)]\n",
    "    \n",
    "    df[df <=2 ] = 0\n",
    "\n",
    "    df=df.T\n",
    "    df=df.loc[np.arange(start,end).astype(str)]\n",
    "    df=df.T\n",
    "\n",
    "    if fp_size_min:\n",
    "        df=df.where(df > fp_size_min, 0)\n",
    "    if fp_size_max:\n",
    "        df=df.where(df < fp_size_max, 0)\n",
    "    df=df.where(df == 0, 1)    \n",
    "    \n",
    "    df_a=df.sum(axis=0)\n",
    "    df_a=np.array(df_a)/counts.to_numpy()\n",
    "    df=''\n",
    "    \n",
    "    return df_a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a function to read in the data\n",
    "def read_data(file_paths, s, e):\n",
    "    df_dict = {}\n",
    "    for path in file_paths:\n",
    "        # read in the path as a dataframe\n",
    "        df = pd.read_csv(path, sep='\\t', names=np.arange(s,e).astype(str))\n",
    "        # extract df_name\n",
    "        df_name = f'{path.split(\"/\")[7]}_{path.split(\"/\")[9].split(\"_\")[0]}'\n",
    "        # store the dataframe in the dictionary with the df_name as the key\n",
    "        df_dict[df_name] = df\n",
    "    return df_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c748bc2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import the data\n",
    "#you can import the entire dataset by leaving out the columns argument or you can choose a position range (s and e)\n",
    "#parquet files require strings as their column names so you have to turn the arange into strings\n",
    "#I would import at minimum 100 or so bases beyond the range you want to work with, just so that the smoothing for some of these plots doesn't have issues\n",
    "# pEF-GFP\n",
    "# s=0\n",
    "# e=5054\n",
    "\n",
    "# Define the base path to directory\n",
    "base_path = '/gscratch/stergachislab/bmallo/large_home/hmm-footprint-caller/pEF_GFP_Gus/'\n",
    "\n",
    "# Define the pattern to match the files\n",
    "# pattern_WT = f'{base_path}Gus_WT*/Outfiles/*_nucleosomes.m6a.gz/*.tsv'\n",
    "# pattern_Mut = f'{base_path}Gus_MUT*/Outfiles/*_nucleosomes.m6a.gz/*.tsv'\n",
    "pattern_GFP = f'{base_path}pEF_GFP_rep3/Outfiles/*_nucleosomes.m6a.gz/*.tsv'\n",
    "\n",
    "# Use glob to get a list of file paths\n",
    "# WT_paths = glob(pattern_WT)\n",
    "# Mut_paths = glob(pattern_Mut)\n",
    "GFP_paths = glob(pattern_GFP)\n",
    "\n",
    "# Create a dictionary to store dataframes in\n",
    "# SLC39A4_WT = read_data(WT_paths, 0, 6462)\n",
    "# SLC39A4_Mut = read_data(Mut_paths, 0, 6462)\n",
    "pEF_GFP = read_data(GFP_paths, 0, 5054)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f2014306",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_dataframes(df_dict, threshold):\n",
    "    df_dict_filtered = {}\n",
    "    for df_name, df in df_dict.items():\n",
    "        # Filter the dataframe to remove all reads whose first footprint is greater than threshold\n",
    "        filtered_df = df[(abs(df.iloc[:, 1]) <= threshold) & (abs(df.iloc[:, 0]) <= threshold)]\n",
    "\n",
    "        # Store the filtered dataframe back in the dictionary\n",
    "        df_dict_filtered[df_name] = filtered_df\n",
    "    return df_dict\n",
    "\n",
    "# Filter dataframes\n",
    "# SLC39A4_WT_filtered = filter_dataframes(SLC39A4_WT, 1000)\n",
    "# SLC39A4_Mut_filtered = filter_dataframes(SLC39A4_Mut, 1000)\n",
    "pEF_GFP_filtered = filter_dataframes(pEF_GFP, 1000)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "aa31bd09",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 's' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m min_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m40\u001b[39m\n\u001b[1;32m      4\u001b[0m max_size\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m250\u001b[39m\n\u001b[0;32m----> 5\u001b[0m start\u001b[38;5;241m=\u001b[39m\u001b[43ms\u001b[49m\n\u001b[1;32m      6\u001b[0m end\u001b[38;5;241m=\u001b[39me\n\u001b[1;32m      8\u001b[0m meta\u001b[38;5;241m=\u001b[39mplot_meta_heatmap(Gus_WT_Rep2_filterd, start, end, min_size, max_size, bin_size)\n",
      "\u001b[0;31mNameError\u001b[0m: name 's' is not defined"
     ]
    }
   ],
   "source": [
    "#plot the footprint size heatmap\n",
    "bin_size=2\n",
    "min_size=40\n",
    "max_size=250\n",
    "start=s\n",
    "end=e\n",
    "\n",
    "meta=plot_meta_heatmap(Gus_WT_Rep2_filterd, start, end, min_size, max_size, bin_size)\n",
    "\n",
    "fig, ax=plt.subplots(1,1,figsize=(10,5))\n",
    "sns.heatmap(data=meta, cbar=True, cmap='gray', ax=ax, xticklabels=1000, yticklabels=10)\n",
    "ax.set_xlabel('genome position')\n",
    "ax.set_ylabel('Size of Footprint (5bp bins)')\n",
    "plt.title('Gus WT nuclear localized plasmid footprints')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "2fc61562",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'pEF_GFP_rep3_35'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[57], line 19\u001b[0m\n\u001b[1;32m     15\u001b[0m PIC_range \u001b[38;5;241m=\u001b[39m (\u001b[38;5;241m60\u001b[39m, \u001b[38;5;241m80\u001b[39m)\n\u001b[1;32m     17\u001b[0m \u001b[38;5;66;03m# sample_1 = SLC39A4_WT_filtered[f'Gus_{allele_1}_{replicate_1}_{count_1}']\u001b[39;00m\n\u001b[1;32m     18\u001b[0m \u001b[38;5;66;03m# sample_2 = SLC39A4_Mut_filtered[f'Gus_{allele_2}_{replicate_2}_{count_2}']\u001b[39;00m\n\u001b[0;32m---> 19\u001b[0m sample_3 \u001b[38;5;241m=\u001b[39m \u001b[43mpEF_GFP_filtered\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43mf\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mpEF_GFP_rep3_\u001b[39;49m\u001b[38;5;132;43;01m{\u001b[39;49;00m\u001b[43mcount_1\u001b[49m\u001b[38;5;132;43;01m}\u001b[39;49;00m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m\n\u001b[1;32m     21\u001b[0m fig, ax \u001b[38;5;241m=\u001b[39m plt\u001b[38;5;241m.\u001b[39msubplots(\u001b[38;5;241m1\u001b[39m,\u001b[38;5;241m1\u001b[39m,figsize\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m10\u001b[39m,\u001b[38;5;241m5\u001b[39m))\n\u001b[1;32m     23\u001b[0m \u001b[38;5;66;03m#grab the enrichment for the different size ranges (can adjust these as you need)\u001b[39;00m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'pEF_GFP_rep3_35'"
     ]
    }
   ],
   "source": [
    "#plot the enrichment of PIC, pause, nucleosome sized footprints\n",
    "s=25\n",
    "e=5029\n",
    "smooth=25\n",
    "i=0\n",
    "\n",
    "# Define samples to plot\n",
    "# allele_1 = 'WT'\n",
    "# allele_2 = 'MUT'    \n",
    "count_1 = 35\n",
    "# count_2 = 3\n",
    "# replicate_1 = 'rep3'\n",
    "# replicate_2 = 'rep3'\n",
    "nuc_range = (90, 1000)\n",
    "PIC_range = (60, 80)\n",
    "\n",
    "# sample_1 = SLC39A4_WT_filtered[f'Gus_{allele_1}_{replicate_1}_{count_1}']\n",
    "# sample_2 = SLC39A4_Mut_filtered[f'Gus_{allele_2}_{replicate_2}_{count_2}']\n",
    "sample_3 = pEF_GFP_filtered[f'pEF_GFP_rep3_{count_1}']\n",
    "\n",
    "fig, ax = plt.subplots(1,1,figsize=(10,5))\n",
    "\n",
    "#grab the enrichment for the different size ranges (can adjust these as you need)\n",
    "a=meta_line_plot(sample_3,  s-smooth, e+smooth, nuc_range[0], nuc_range[1])\n",
    "# b=meta_line_plot(sample_2,  s-smooth, e+smooth, nuc_range[0], nuc_range[1])\n",
    "c=meta_line_plot(sample_3,  s-smooth, e+smooth, PIC_range[0], PIC_range[1])\n",
    "# d=meta_line_plot(sample_2,  s-smooth, e+smooth, PIC_range[0], PIC_range[1])\n",
    "\n",
    "#do some smoothing to make it look pretty\n",
    "box = np.ones(smooth)/smooth\n",
    "a = np.convolve(a, box, mode='same')\n",
    "a = a[smooth:len(a)-smooth]\n",
    "# b = np.convolve(b, box, mode='same')\n",
    "# b = b[smooth:len(b)-smooth]\n",
    "c = np.convolve(c, box, mode='same')\n",
    "c = c[smooth:len(c)-smooth]\n",
    "# d = np.convolve(d, box, mode='same')\n",
    "# d = d[smooth:len(d)-smooth]\n",
    "\n",
    "\n",
    "# plot PIC/Pol II\n",
    "ax.plot(np.arange(s,e),c, color='#009688', linestyle='solid', linewidth=1.5, label=f'PIC')\n",
    "# ax.plot(np.arange(s,e),d, color='#F57C00', linestyle='solid', linewidth=1.5, label=f'{count_2}')\n",
    "ax.set_ylim(0, 0.1)\n",
    "ax.set_xlim(s,e)\n",
    "\n",
    "# Add vertical line at SNP position\n",
    "#vline = ax.axvline(x=6119, color='black', linestyle='dotted', linewidth=1.4, label='A>G')\n",
    "\n",
    "#plot nucleosomes on a twin axis bc there are way more\n",
    "ax2 = ax.twinx() \n",
    "#ax2.set_xlim(s,e)\n",
    "#ax2.set_ylim(0, 1)\n",
    "# ax2.plot(np.arange(s,e),a, color='#009688', label=f'{allele_1}: {count_1} nucs')\n",
    "ax2.plot(np.arange(s,e),a, color='#F57C00', label=f'{count_1} nucs')\n",
    "# ax2.plot(np.arange(s,e),b, color='#F57C00', label=f'{allele_2}: {count_2} nucs')\n",
    "\n",
    "# Combine legend handles and labels\n",
    "# lines, labels = ax.get_legend_handles_labels()\n",
    "# lines2, labels2 = ax2.get_legend_handles_labels()\n",
    "# lines2.append(lines[2]) # add the nucleosome lines to the legend\n",
    "# labels2.append(labels[2])\n",
    "\n",
    "# Add legend manually\n",
    "ax.legend(loc='upper right', frameon=False, bbox_to_anchor=(1.05, 1.2), fontsize=10)\n",
    "ax.set_ylim(0, 0.15)\n",
    "ax2.set_ylim(0, 1)\n",
    "\n",
    "# Set labels and title \n",
    "plt.xlabel('genomic position')\n",
    "ax.set_ylabel('footprint enrichment (Pol II)')\n",
    "ax2.set_ylabel('footprint enrichment (Nucleosomes)')\n",
    "plt.title(f'pEF-GFP Rep3: {count_1}')\n",
    "plt.legend(loc='upper right', frameon=False, bbox_to_anchor=(1.09, 1.15), fontsize=10)\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save figure\n",
    "output_path = f'/gscratch/stergachislab/bmallo/large_home/hmm-footprint-caller/figures/pEF-GFP_rep3/pEF-GFP_rep3_{count_1}_nucs.png'\n",
    "\n",
    "# Add metadata\n",
    "# Set standardized metadata information\n",
    "metadata = {\n",
    "    'Nucleosome Sizes': str(nuc_range),\n",
    "    'PIC sizes': str(PIC_range),  # You can include custom keys as well\n",
    "}\n",
    "\n",
    "# Save figure\n",
    "plt.savefig(output_path, metadata=metadata, bbox_inches='tight', dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_enrichment(s, e, smooth, allele_1, allele_2, count_1, count_2, replicate_1, replicate_2, nuc_range=(90,10000), PIC_range=(60, 80)):\n",
    "\n",
    "    sample_1 = SLC39A4_WT_filtered[f'Gus_{allele_1}_{replicate_1}_{count_1}']\n",
    "    sample_2 = SLC39A4_Mut_filtered[f'Gus_{allele_2}_{replicate_2}_{count_2}']\n",
    "\n",
    "    fig, ax = plt.subplots(1, 1, figsize=(10, 5))\n",
    "\n",
    "    a = meta_line_plot(sample_1, s - smooth, e + smooth, nuc_range[0], nuc_range[1])\n",
    "    b = meta_line_plot(sample_2, s - smooth, e + smooth, nuc_range[0], nuc_range[1])\n",
    "    c = meta_line_plot(sample_1, s - smooth, e + smooth, PIC_range[0], PIC_range[1])\n",
    "    d = meta_line_plot(sample_2, s - smooth, e + smooth, PIC_range[0], PIC_range[1])\n",
    "\n",
    "    box = np.ones(smooth) / smooth\n",
    "    a = np.convolve(a, box, mode='same')[smooth:len(a) - smooth]\n",
    "    b = np.convolve(b, box, mode='same')[smooth:len(b) - smooth]\n",
    "    c = np.convolve(c, box, mode='same')[smooth:len(c) - smooth]\n",
    "    d = np.convolve(d, box, mode='same')[smooth:len(d) - smooth]\n",
    "\n",
    "    ax.plot(np.arange(s,e),c, color='#009688', linestyle='dashed', linewidth=1.5, label=f'{count_1}')\n",
    "    ax.plot(np.arange(s,e),d, color='#F57C00', linestyle='dashed', linewidth=1.5, label=f'{count_2}')\n",
    "    ax.set_ylim(0, 0.1)\n",
    "    ax.set_xlim(s, e)\n",
    "\n",
    "    vline = ax.axvline(x=6119, color='black', linestyle='dotted', linewidth=1.4, label='A>G')\n",
    "    \n",
    "    ax2 = ax.twinx()\n",
    "    ax2.set_ylim(0, 1)\n",
    "    ax2.plot(np.arange(s,e),a, color='#009688', label=f'{allele_1}: {count_1} nucs')\n",
    "    ax2.plot(np.arange(s,e),b, color='#F57C00', label=f'{allele_2}: {count_2} nucs')\n",
    "\n",
    "    lines, labels = ax.get_legend_handles_labels()\n",
    "    lines2, labels2 = ax2.get_legend_handles_labels()\n",
    "    lines2.append(lines[2])\n",
    "    labels2.append(labels[2])\n",
    "\n",
    "    ax.legend(lines2, labels2, loc='upper right', frameon=False, bbox_to_anchor=(1.05, 1.2), fontsize=10)\n",
    "\n",
    "    plt.xlabel('genomic position')\n",
    "    ax.set_ylabel('footprint enrichment (Pol II)')\n",
    "    ax2.set_ylabel('footprint enrichment (Nucleosomes)')\n",
    "    plt.title(f'SLC39A4: WT ({replicate_1}) vs MUT ({replicate_2})')\n",
    "    plt.tight_layout()\n",
    "\n",
    "    output_path = f'/gscratch/stergachislab/bmallo/large_home/hmm-footprint-caller/figures/SLC39A4_WT_vs_MUT_2/SLC39A4_WT_vs_MUT_{replicate_1}_vs_{replicate_2}_{count_1}_nucs.png'\n",
    "\n",
    "    metadata = {\n",
    "        'Nucleosome Sizes': str(nuc_range),\n",
    "        'PIC sizes': str(PIC_range),\n",
    "    }\n",
    "\n",
    "    plt.savefig(output_path, metadata=metadata, bbox_inches='tight', dpi=300)\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample not found for count=40 and replicate=rep1. Skipping...\n",
      "Sample not found for count=41 and replicate=rep1. Skipping...\n",
      "Sample not found for count=42 and replicate=rep1. Skipping...\n",
      "Sample not found for count=42 and replicate=rep2. Skipping...\n",
      "Sample not found for count=43 and replicate=rep1. Skipping...\n",
      "Sample not found for count=43 and replicate=rep2. Skipping...\n",
      "Sample not found for count=43 and replicate=rep3. Skipping...\n",
      "Sample not found for count=44 and replicate=rep1. Skipping...\n",
      "Sample not found for count=44 and replicate=rep2. Skipping...\n",
      "Sample not found for count=44 and replicate=rep3. Skipping...\n"
     ]
    }
   ],
   "source": [
    "for count in range(1, 45):\n",
    "    for replicate in ['rep1', 'rep2', 'rep3']:\n",
    "        count_1 = count_2 = count\n",
    "        replicate_1 = replicate_2 = replicate\n",
    "\n",
    "        try:\n",
    "            plot_enrichment(s=25, e=6437, smooth=25, allele_1='WT', allele_2='MUT',\n",
    "                            count_1=count_1, count_2=count_2, replicate_1=replicate_1, replicate_2=replicate_2)\n",
    "        except KeyError:\n",
    "            # Handle the case where the sample does not exist\n",
    "            print(f\"Sample not found for count={count} and replicate={replicate}. Skipping...\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e78d936",
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot all reads in a given region, colored by footprint identity\n",
    "\n",
    "s=25\n",
    "e=6400\n",
    "\n",
    "fig, ax=plt.subplots(1,1, figsize=(10,30))\n",
    "\n",
    "d=SLC39A4_WT_filtered['Gus_WT_rep3_22']\n",
    "#you can do stuff like k-means clustering the dataframe or sorting by the distance to the 1st nucleosome or something to get different orders of reads if you want\n",
    "\n",
    "# Specify the number of rows and the column index to center on\n",
    "num_rows_to_plot = 400\n",
    "center_column_index = 6000  # Replace with the desired column index\n",
    "\n",
    "#these plot the white space and black lines\n",
    "for i in range(d.shape[0]+1):\n",
    "    ax.axhline(i+.5, color='black', lw=.25, zorder=1)\n",
    "for i in range(d.shape[0]+1):\n",
    "    ax.axhline(i, color='white', lw=2, zorder=3)\n",
    "\n",
    "#this plots the actual reads\n",
    "sns.heatmap(data=d,\n",
    "cmap=pol_cp, vmin=0, vmax=1000,\n",
    "            yticklabels=False, cbar=False, ax=ax,\n",
    "            xticklabels=500)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
